File: Leetcode_POTD_2025/8_August/1504_Count_Submatrices_With_All_Ones.cpp
class Solution {
public:
    int numSubmat(std::vector<std::vector<int>>& mat) {
        int rows = mat.size();
        int cols = mat[0].size();
        std::vector<std::vector<int>> heights(rows, std::vector<int>(cols, 0));
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (mat[i][j] == 1) {
                    heights[i][j] = (i > 0) ? heights[i - 1][j] + 1 : 1;
                }
            }
        }
        int total = 0;
        for (int i = 0; i < rows; i++) {
            std::stack<std::vector<int>> st;
            int count = 0;
            for (int j = 0; j < cols; j++) {
                int h = heights[i][j];
                while (!st.empty() && st.top()[0] > h) {
                    st.pop();
                }
                int prev_count = st.empty() ? 0 : st.top()[1];
                int current_count = h * (j - (st.empty() ? -1 : st.top()[2])) + prev_count;
                total += current_count;
                st.push({h, current_count, j});
            }
        }
        return total;
    }
};

File: Leetcode_POTD_2025/8_August/1504_Count_Submatrices_With_All_Ones.py
class Solution(object):
    def numSubmat(self, mat):
        rows = len(mat)
        cols = len(mat[0])
        heights = [[0] * cols for _ in range(rows)]
        for i in range(rows):
            for j in range(cols):
                if mat[i][j] == 1:
                    heights[i][j] = heights[i - 1][j] + 1 if i > 0 else 1
        total = 0
        for i in range(rows):
            stack = []
            count = 0
            for j in range(cols):
                h = heights[i][j]
                while stack and stack[-1][0] > h:
                    stack.pop()
                prev_count = stack[-1][1] if stack else 0
                current_count = h * (j - (stack[-1][2] if stack else -1)) + prev_count
                total += current_count
                stack.append([h, current_count, j])
        return total
